Extend Ivy Functional API  huber_loss

sushmanthreddy opened the pull request:
close #21069

hello-fri-end reviewed (changes requested):
Hey @sushmanthreddy ! Thanks for opening this PR! Would you please add a container and array method  in `ivy/container/experimental/losses.py` and the array instance method in `ivy/array/experimental/losses.py`. Feel free to request another review when you are done.

sushmanthreddy commented:
> Hey @sushmanthreddy ! Thanks for opening this PR! Would you please add a container and array method in `ivy/data-classes/container/experimental/losses.py` and `ivy/data-classes/array/experimental/losses.py` respectively. Feel free to request another review when you are done.

i have added the container and array static methods..
pls once review it and merge it

hello-fri-end commented on ivy/functional/ivy/losses.py:
```diff
@@ -383,3 +383,56 @@ def sparse_cross_entropy(
+    >>> huber_loss(y_true, y_pred, delta=0.5)
+    ivy.array([0.25 , 0.25 , 0.   , 0.125])
+    """
+    ivy.utils.assertions.assert_float_param(delta, 'delta', allow_inf=False)
```
There's no function `assert_float_param` in `assertions`. I am not sure what exactly you are trying to do here.  Could you please explain this?

hello-fri-end commented:
ivy-gardener

sushmanthreddy commented:
@hello-fri-end can u pls review the pr ?i have made necessary changes

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,70 @@ def log_poisson_loss(
+    y_true: Union[ivy.Array, ivy.NativeArray],
+    y_pred: Union[ivy.Array, ivy.NativeArray],
+    /,
+    delta: float = 1.0,
+    reduction: str = "mean",
```
```suggestion
    delta: Optional[float] = 1.0,
    reduction: Optional[str] = "mean",
```

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,70 @@ def log_poisson_loss(
+    >>> huber_loss(y_true, y_pred, delta=0.5)
+    ivy.array([0.25 , 0.25 , 0.   , 0.125])
+    """
+    assert isinstance(delta, float), "The 'delta' parameter must be a float."
```
Tbh, I don't think there's any need to add this since we have already mentioned in the type-hints that `delta` should be a float.

sushmanthreddy replied:
yeah i thought of mentioning that it should be float only, anyways removed it..

hello-fri-end commented on ivy/data_classes/array/experimental/losses.py:
```diff
@@ -1,6 +1,48 @@
+class _ArrayWithLosses(abc.ABC):
+    def huber_loss(
+        self: ivy.Array,
+        y_pred: Union[ivy.Array, ivy.NativeArray],
```
``` suggestion
       pred: Union[ivy.Array, ivy.NativeArray],
```

Just to be consistent.

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,70 @@ def log_poisson_loss(
+@inputs_to_ivy_arrays
+@handle_array_function
+def huber_loss(
+    y_true: Union[ivy.Array, ivy.NativeArray],
+    y_pred: Union[ivy.Array, ivy.NativeArray],
```
```suggestion
    true: Union[ivy.Array, ivy.NativeArray],
    pred: Union[ivy.Array, ivy.NativeArray],
```
To be consistent with other loss functions.

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,70 @@ def log_poisson_loss(
+
+    >>> huber_loss(y_true, y_pred, delta=0.5)
+    ivy.array([0.25 , 0.25 , 0.   , 0.125])
+    """
```
```suggestion
        try:
        assert true.shape == pred.shape
    except ValueError:
        raise ValueError(
            "`pred` and `true` must have the same shape, received "
            f"({pred.shape} vs {true.shape})."
        )
```

We should be asserting that the shapes are equal.

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,70 @@ def log_poisson_loss(
+
+    Parameters
+    ----------
+    y_true : array_like
+        The true (ground truth) values.
+    y_pred : array_like
```
We should changes these to `true` and `pred` as well.

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,70 @@ def log_poisson_loss(
+    loss = ivy.where(abs_diff <= delta, quadratic_loss, linear_loss)
+
+    if reduction == "mean":
+        ret = ivy.mean(loss)
```
```suggestion
        ret = ivy.mean(loss, out=out)
```

So that the `out` parameter is also updated.

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,70 @@ def log_poisson_loss(
+    if reduction == "mean":
+        ret = ivy.mean(loss)
+    elif reduction == "sum":
+        ret = ivy.sum(loss)
```
```suggestion
        ret = ivy.sum(loss, out=out)
```

hello-fri-end reviewed (changes requested):
Hey @sushmanthreddy ! Left a few more comments :)

Also, since pytorch has a functional implementation for `huber_loss` (https://pytorch.org/docs/stable/generated/torch.nn.functional.huber_loss.html), this should be implemented as a [mixed function](https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions) and you should a add primary implementation for torch backend in `ivy/functional/backends/torch/experimental/losses.py`. Let me know if I should clarify anything further. Feel free to request another review when you are done with the changes, Thanks and happy coding :)

sushmanthreddy commented:
@hello-fri-end sorry for so many commits , i have changed it on locally,


are there any required changes..to be made?

hello-fri-end commented on ivy/data_classes/array/experimental/losses.py:
```diff
@@ -1,6 +1,48 @@
+        self: ivy.Array,
+       pred: Union[ivy.Array, ivy.NativeArray],
+        /,
+        delta: float = 1.0,
```
```suggestion
        delta: Optional[float] = 1.0,
```

hello-fri-end commented on ivy/data_classes/array/experimental/losses.py:
```diff
@@ -1,6 +1,48 @@
+        ----------
+        self
+            The true (ground truth) values.
+        y_pred
```
```suggestion
        pred
```

hello-fri-end commented on ivy/data_classes/array/experimental/losses.py:
```diff
@@ -1,6 +1,48 @@
+
+        Examples
+        --------
+        >>> y_true = ivy.array([2, 4, 7, 1])
+        >>> y_pred = ivy.array([2.5, 3.5, 8, 0.8])
+        >>> loss = y_true.huber_loss(y_pred, delta=1.0)
```
let's change these to `true` and `pred`

hello-fri-end commented on ivy/data_classes/array/experimental/losses.py:
```diff
@@ -1,6 +1,48 @@
+        >>> print(loss)
+        ivy.array([0.125, 0.125, 0.5  , 0.125])
+        """
+        return ivy.huber_loss(self._data, y_pred, delta=delta, out=out)
```
```suggestion
        return ivy.huber_loss(self._data, pred, delta=delta, out=out)
```

hello-fri-end commented on ivy/data_classes/container/experimental/losses.py:
```diff
@@ -0,0 +1,162 @@
+        pred: Union[ivy.Container, ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        delta: Union[float, ivy.Container] = 1.0,
```
```suggestion
        delta: Union[float] = 1.0,
```

Only arrays arguments can be containers.

hello-fri-end commented on ivy/data_classes/container/experimental/losses.py:
```diff
@@ -0,0 +1,162 @@
+        pred: Union[ivy.Container, ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        delta: Union[float, ivy.Container] = 1.0,
```
```suggestion
        delta: Union[float] = 1.0,
```

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,77 @@ def log_poisson_loss(
+            "`pred` and `true` must have the same shape, received "
+            f"({pred.shape} vs {true.shape})."
+        )
+    abs_diff = ivy.abs(y_true - y_pred)
```
```suggestion
    abs_diff = ivy.abs(true - pred)
```

hello-fri-end reviewed (changes requested):
Have left a few more comments for you to resolve. Also, can you add a backend implementation for torch as I mentioned in the last review. Feel free to request another review when you are done. Thanks :)

sushmanthreddy commented:
@hello-fri-end where , should i add in torch module there is no file called losses in torch..

I am new to ivy , I really don't knew about torch backend implementation, if you can show me a little example ,I can replicate that for huberloss

hello-fri-end commented:
> @hello-fri-end where , should i add in torch module there is no file called losses in torch..
The function needs to be added here :`ivy/functional/backends/torch/experimental/losses.py`. You can create the  `losses.py` file if it does not already exist.

> I am new to ivy , I really don't knew about torch backend implementation, if you can show me a little example ,I can replicate that for huberloss
You can take a look at https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions for understanding mixed functions. To summarise though, the function signature in the torch backend implementation should be the same as in the functional implementation and you should use the `torch.nn.functional.huber_loss` function in the implementation. Let me know if you need me to clarify anything further. Thanks :)

sushmanthreddy commented:
@hello-fri-end loss functions are composite functions . so it doesn't need backend implementation, I think we can implement in stateful function.
<img width="804" alt="Screenshot 2023-08-03 at 10 55 51 AM" src="https://github.com/unifyai/ivy/assets/73489688/9cb4de80-1a45-4d34-b4e0-71549c1cbfb7">

hello-fri-end commented:
> @hello-fri-end loss functions are composite functions . so it doesn't need backend implementation, I think we can implement in stateful function. <img alt="Screenshot 2023-08-03 at 10 55 51 AM" width="804" src="https://user-images.githubusercontent.com/73489688/258008645-9cb4de80-1a45-4d34-b4e0-71549c1cbfb7.png">

If there exists a native implementation for a function within any framework, we always add it to our backends. This is because the native implementations are mostly written in C++ which makes them more efficient than having a compositional implementation in Ivy. In this case, since torch has a `huber_loss` function, we should add it in our torch backend implementations. Hope that makes sense. Also, please go through the deep-dive I shared above for more details.

sushmanthreddy commented:
@hello-fri-end backend in torch is implemented 
i have used l1_loss

hello-fri-end commented:
> @hello-fri-end backend in torch is implemented i have used l1_loss
> 
> @hello-fri-end pls give a look on this reply once [link](https://github.com/unifyai/ivy/pull/21070#discussion_r1282567900)

You should implement huber_loss in the backend and not l1_loss.

sushmanthreddy commented:
@hello-fri-end sorry i was working on l1_loss also..got confused without seeing

sushmanthreddy commented:
@hello-fri-end sorry for the mistake i have use torch.nn.HuberLoss function and updated the backend implementation

sushmanthreddy commented:
@hello-fri-end can u pls review again?I have added the huber loss backen implementation in torch

hello-fri-end commented on ivy/functional/backends/torch/experimental/losses.py:
```diff
@@ -0,0 +1,22 @@
+
+# global
+import torch
+import torch.nn
```
Let's remove this since it's not used.
```suggestion
```

hello-fri-end commented on ivy/functional/backends/torch/experimental/losses.py:
```diff
@@ -0,0 +1,22 @@
+    reduction: str = "mean",
+    out: torch.Tensor = None,
+) -> torch.Tensor:
+    huber_loss_fn = torch.nn.HuberLoss(reduction=reduction, delta=delta)
```
You should be using `torch.nn.functional.huber_loss` here.Look at the documentation here https://pytorch.org/docs/stable/generated/torch.nn.functional.huber_loss.html and update the function call accordingly.

hello-fri-end commented on ivy/functional/backends/torch/experimental/losses.py:
```diff
@@ -0,0 +1,22 @@
+
+
+
+@with_unsupported_dtypes({"2.0.1 and below": "float"}, backend_version)
```
You should run this script to check data-types supported by this function:
```python
import torch

def get_supported_dtypes():
    supported_dtypes = []
    unsupported_dtypes = []
    
    # Define all supported data types
    data_types = [torch.float16, torch.float32, torch.float64,
                  torch.bfloat16, torch.uint8, torch.int8,
                  torch.int16, torch.int32, torch.int64, torch.bool]
    
    for dtype in data_types:
        try:
            # Sample input tensors
            y_true = torch.tensor([1.0, 2.0, 3.0], dtype=dtype)
            y_pred = torch.tensor([1.1, 1.8, 2.9], dtype=dtype)
            
            # Check if the function raises an exception
            torch.nn.functional.huber_loss(y_true, y_pred)
            
            supported_dtypes.append(dtype)
        except :
            unsupported_dtypes.append(dtype)
    
    return supported_dtypes, unsupported_dtypes

if __name__ == "__main__":
    supported_dtypes, unsupported_dtypes = get_supported_dtypes()
    
    print("Supported dtypes:", supported_dtypes)
    print("Unsupported dtypes:", unsupported_dtypes)
   ```

It will return 

```python
Supported dtypes: [torch.float16, torch.float32, torch.float64, torch.bfloat16]
Unsupported dtypes: [torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64, torch.bool]
```

Please update the decorator accordingly.

hello-fri-end reviewed (changes requested):
Please resolve these comments. Make sure that the tests are passing locally first. Feel free to request another review when you are done. Happy coding :)

hello-fri-end commented on ivy/functional/ivy/losses.py:
```diff
@@ -383,3 +383,5 @@ def sparse_cross_entropy(
     return ivy.cross_entropy(
         true, pred, axis=axis, epsilon=epsilon, reduction=reduction, out=out
     )
+
```
```suggestion
huber_loss.mixed_backend_wrappers = {
    "to_add": (
        "handle_out_argument",
        "inputs_to_native_arrays",
        "outputs_to_ivy_arrays",
        "handle_device_shifting",
    ),
    "to_skip": ("inputs_to_ivy_arrays"),
}
```
Since it's a mixed function you need to add this.

sushmanthreddy commented:
@hello-fri-end I am still confused ,why u mentioned me to write huberloss as mixed function ,because it is only supported by torch and tensor flow ...same like binary cross entropy and backend is not implemented for binary cross entrpy loss..

so , by taking example have implemented huber loss composite function.

Now test cases will pass ,only for paddle they will not  pass because of bfloat bug.
And u can see the bfloat bug here in this discussion. [link](https://discord.com/channels/799879767196958751/1139168214723592264)


And stateful API also has bug due to that we get an error `modulenotfound`.
And tests will not pass not only for huber loss which is implemented above but also for sparse_cross_entropy .

sushmanthreddy commented:
@hello-fri-end I have implemented the huber_loss for above function and now tests are passing ,i have added the backend support also,can u pls review the pull request?

The paddle test is failing because of an error with the bfloat16 datatype which again is a known issue that Ivy people are working on.

so pls review it once again.

i hope will get those magical words "LGTM"..lol

sushmanthreddy commented:
![Screenshot from 2023-08-18 02-25-37](https://github.com/unifyai/ivy/assets/73489688/a4913144-9d14-4a04-8830-d5ca20b6851f)
tests passing sc

hello-fri-end commented on ivy/data_classes/array/experimental/losses.py:
```diff
@@ -1,6 +1,52 @@
+        pred: Union[ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        reduction: str = "mean",
```
```suggestion
        reduction: Optional[str] = "mean",
```

I believe this parameter should also be optional.

hello-fri-end commented on ivy/data_classes/array/experimental/losses.py:
```diff
@@ -1,6 +1,52 @@
+            The predicted values by the model.
+        delta
+            The threshold parameter that determines the point where the loss transitions
+            from squared error to absolute error. Default is 1.0.
```
Can you add `reduction` parameter here also?

hello-fri-end commented on ivy/data_classes/container/experimental/losses.py:
```diff
@@ -0,0 +1,166 @@
+        pred: Union[ivy.Container, ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        delta: Union[float, ivy.Container] = 1.0,
```
```suggestion
        delta: Optional[Union[float, ivy.Container]] = 1.0,
```

hello-fri-end commented on ivy/data_classes/container/experimental/losses.py:
```diff
@@ -0,0 +1,166 @@
+        /,
+        *,
+        delta: Union[float, ivy.Container] = 1.0,
+        reduction: Union[str, ivy.Container] = "mean",
```
```suggestion
        reduction: Optional[Union[str, ivy.Container]] = "mean",
```

hello-fri-end commented on ivy/data_classes/container/experimental/losses.py:
```diff
@@ -0,0 +1,166 @@
+            input array or container containing the predicted labels.
+        delta
+            The threshold parameter that determines the point where the loss transitions
+            from squared error to absolute error. Default is 1.0.
```
`reduction` here aswell./

hello-fri-end commented on ivy/data_classes/container/experimental/losses.py:
```diff
@@ -0,0 +1,166 @@
+        /,
+        *,
+        delta: Union[float, ivy.Container] = 1.0,
+        reduction: Union[str, ivy.Container] = "mean",
```
```suggestion
        reduction: Optional[Union[str, ivy.Container]] = "mean",
```

hello-fri-end commented on ivy/data_classes/container/experimental/losses.py:
```diff
@@ -0,0 +1,166 @@
+        pred: Union[ivy.Container, ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        delta: Union[float, ivy.Container] = 1.0,
```
```suggestion
        delta: Optional[Union[float, ivy.Container]] = 1.0,
```

hello-fri-end commented on ivy/data_classes/container/experimental/losses.py:
```diff
@@ -0,0 +1,166 @@
+        delta
+            The threshold parameter that determines the point where the loss transitions
+            from squared error to absolute error. Default is 1.0.
+        key_chains
```
`reduction` here.

hello-fri-end commented on ivy/functional/backends/paddle/experimental/losses.py:
```diff
@@ -0,0 +1,35 @@
+    backend_version,
+)
+def huber_loss(
+    input: paddle.Tensor,
+    label: paddle.Tensor,
+    /,
+    *,
+    delta: Optional[float] = 1.0,
+    out: Optional[paddle.Tensor] = None,
+) -> paddle.Tensor:
+    return paddle.fluid.layers.huber_loss(input, label, delta=delta, out=out)
```
The function signature should match in all the implementations.

```python
def huber_loss(
    true: Union[paddle.tensor],
    pred: Union[paddle.tensor],
    /,
    *,
    delta: Optional[float] = 1.0,
    reduction: Optional[str] = "mean",
    out: Optionalpaddle.tensor] = None,
```

sushmanthreddy replied:
@hello-fri-end can u plss see function implementation here , in huber loss and it doesn't have the reduction parameter in its backend ,that is the reason ,I have left out reduction over here..

<img width="1101" alt="Screenshot 2023-08-19 at 11 27 45 AM" src="https://github.com/unifyai/ivy/assets/73489688/cdbf21b2-3db4-4d37-9040-afcd60d8a183">

hello-fri-end commented on ivy/functional/backends/torch/experimental/losses.py:
```diff
@@ -0,0 +1,21 @@
+
+@with_unsupported_dtypes({"2.0.1 and below": ("float16",)}, backend_version)
+def huber_loss(
+    input: torch.Tensor,
+    target: torch.Tensor,
```
Again, the function signature should match. These should be renamed to `true` and `pred`

hello-fri-end commented on ivy/functional/ivy/experimental/losses.py:
```diff
@@ -93,3 +94,68 @@ def log_poisson_loss(
+    linear_loss = delta * (abs_diff - 0.5 * delta)
+    loss = ivy.where(abs_diff <= delta, quadratic_loss, linear_loss)
+
+    if reduction == "sum":
+        return ivy.sum(loss, out=out)
+    elif reduction == "mean":
+        return ivy.mean(loss, out=out)
+    else:
+        return ivy.inplace_update(out, loss) if out is not None else loss
```
I'm not sure if I follow this completely. `reduction` can only have two values `sum` or `mean`, so, why is there an else case here?

sushmanthreddy replied:
@hello-fri-end here the case where it is "none " is there actually none is also there for them implementation i that case else will work

sushmanthreddy commented:
@hello-fri-end I have made necessary changes ,can u pls review it once again?if no changes are needed pls merge it..
